File: android/src/main/java/com/lynkxyz/noke/RNNokeModule.java
package com.lynkxyz.noke;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.IBinder;
import android.support.annotation.Nullable;
import android.system.ErrnoException;
import android.util.Log;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import com.facebook.react.uimanager.IllegalViewOperationException;
import com.noke.nokemobilelibrary.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.IllformedLocaleException;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class RNNokeModule extends ReactContextBaseJavaModule {
  public static final String REACT_CLASS = "RNNoke";
  private static ReactApplicationContext reactContext = null;

  public static final String TAG = "RNNoke";
  private NokeDeviceManagerService mNokeService = null;
  private NokeDevice currentNoke;
  private int lastEventCode = -1;

  public RNNokeModule(ReactApplicationContext context) {
    // Pass in the context to the constructor and save it so you can emit events
    // https://facebook.github.io/react-native/docs/native-modules-android.html#the-toast-module
    super(context);

    reactContext = context;
  }

  private WritableMap createCommonEvents(NokeDevice nokeDevice) {
    final WritableMap event = Arguments.createMap();

    if(nokeDevice == null) {
      return event;
    }

    event.putString("name", nokeDevice.getName());
    event.putString("mac", nokeDevice.getMac());
    event.putString("session", nokeDevice.getSession());
    event.putBoolean("status", true);

    return event;
  }

  @ReactMethod
  private void initiateNokeService(Promise promise) {
    try {
      Intent nokeServiceIntent = new Intent(reactContext, NokeDeviceManagerService.class);
      getReactApplicationContext().bindService(nokeServiceIntent, mServiceConnection, Context.BIND_AUTO_CREATE);
//      mNokeService.setUploadUrl("https://v1.api.nokepro.com/lock/upload/");

      WritableMap event = Arguments.createMap();
      event.putBoolean("status", true);

      promise.resolve(event);
    } catch (IllegalViewOperationException e) {
      promise.reject("message", e.getMessage());
    }
  }

  @ReactMethod
  public void startScan(Promise promise) {
    mNokeService.startScanningForNokeDevices();
    final WritableMap event = Arguments.createMap();
    event.putBoolean("status", true);

    promise.resolve(event);
  }

  @ReactMethod
  public void stopScan(Promise promise) {
    mNokeService.stopScanning();
    final WritableMap event = Arguments.createMap();
    event.putBoolean("status", true);

    promise.resolve(event);
  }

  private NokeDevice getCurrentNoke(String macAddress) {
    if(mNokeService != null) {
      LinkedHashMap<String, NokeDevice> nokeDevices = mNokeService.nokeDevices;

      return nokeDevices.get(macAddress);
    }

    return null;
  }

  private NokeDevice addNokeIfNeeded(NokeHashMap nokeHashMap) {
    String name = nokeHashMap.getName();
    String key = nokeHashMap.getKey();
    String command = nokeHashMap.getCommand();
    String macAddress = nokeHashMap.getMacAddress();

    NokeDevice nokeDevice = getCurrentNoke(macAddress);

    if(mNokeService == null) {
      return null;
    }

    if(nokeDevice == null) {
      nokeDevice = new NokeDevice(
              name,
              macAddress
      );
      mNokeService.addNokeDevice(nokeDevice);
    }

    if(nokeDevice.getOfflineUnlockCmd() == null) {
      if(key != null && command != null) {
        nokeDevice.setOfflineUnlockCmd(command);
        nokeDevice.setOfflineKey(key);
      }
    }

    currentNoke = nokeDevice;

    return nokeDevice;
  }

  @ReactMethod
  public void addNokeDevice(ReadableMap data, Promise promise) {
    NokeHashMap nokeHashMap = new NokeHashMap(data);

    NokeDevice nokeDevice = addNokeIfNeeded(nokeHashMap);

    if(nokeDevice == null) {
      promise.reject("300", "Can't get or add nokeDevice");
      return;
    }

    promise.resolve(createCommonEvents(nokeDevice));
  }

  @ReactMethod
  public void addNokeDeviceOnce(ReadableMap data, Promise promise) {
    NokeHashMap nokeHashMap = new NokeHashMap(data);
    String macAddress = nokeHashMap.getMacAddress();

    if(currentNoke != null && !currentNoke.getMac().equals(macAddress)) {
      mNokeService.removeNokeDevice(macAddress);
    }

    NokeDevice nokeDevice = addNokeIfNeeded(nokeHashMap);

    if(nokeDevice == null) {
      promise.reject("300", "Can't get or add nokeDevice");
      return;
    }

    promise.resolve(createCommonEvents(nokeDevice));
  }

  @ReactMethod
  public void sendCommands(ReadableMap data, Promise promise) {
    NokeHashMap nokeHashMap = new NokeHashMap(data);
    String macAddress = nokeHashMap.getMacAddress();
    ArrayList<String> commands = nokeHashMap.getCommands();
    NokeDevice nokeDevice = getCurrentNoke(macAddress);

    if(nokeDevice == null) {
      promise.reject("100", "Noke device is null");
      return;
    }

    if(commands != null) {
      for (int i = 0; i < commands.size(); i++) {
        nokeDevice.sendCommands(commands.get(i));
      }
    }

    promise.resolve(createCommonEvents(nokeDevice));
  }

  @ReactMethod
  public void removeAllNokes(Promise promise) {
    mNokeService.removeAllNoke();
    promise.resolve(null);
  }

  @ReactMethod
  public void removeNokeDevice(String mac, Promise promise) {
    mNokeService.removeNokeDevice(mac);

    promise.resolve(null);
  }

  @ReactMethod
  public void offlineUnlock(ReadableMap data, Promise promise) {
    NokeHashMap nokeHashMap = new NokeHashMap(data);
    String macAddress = nokeHashMap.getMacAddress();
    NokeDevice nokeDevice = getCurrentNoke(macAddress);

    if(nokeDevice == null) {
      promise.reject("100", "Noke device is null");
      return;
    }

    WritableMap event = createCommonEvents(nokeDevice);
    if(lastEventCode == 4) {
      promise.resolve(event);
      return;
    }

    if (nokeDevice.getOfflineUnlockCmd() != null){
      nokeDevice.offlineUnlock();
    }
    promise.resolve(event);
  }

  @ReactMethod
  public void getDeviceInfo(Promise promise) {
    WritableMap event = Arguments.createMap();

    if (currentNoke == null) {
      event.putBoolean("status", false);
    } else {
      event.putBoolean("status", true);
      event.putString("name", currentNoke.getName());
      event.putInt("battery", currentNoke.getBattery());
      event.putString("mac", currentNoke.getMac());
      event.putString("offlineKey", currentNoke.getOfflineKey());
      event.putString("offlineUnlockCmd", currentNoke.getOfflineUnlockCmd());
      event.putString("serial", currentNoke.getSerial());
      event.putString("session", currentNoke.getSession());
      event.putString("trackingKey", currentNoke.getTrackingKey());
      event.putDouble("lastSeen", currentNoke.getLastSeen());
      event.putString("version", currentNoke.getVersion());
    }

    promise.resolve(event);
  }

  @ReactMethod
  public void setAPIKey(String apiKey, Promise promise) {
    mNokeService.setApiKey(apiKey);

    promise.resolve(null);
  }

  @Override
  public String getName() {
    return REACT_CLASS;
  }

  @Override
  public Map<String, Object> getConstants() {
    final Map<String, Object> constants = new HashMap<>();
    constants.put("AUTHOR", "linh_the_human");

    return constants;
  }

  private ServiceConnection mServiceConnection = new ServiceConnection() {
    public void onServiceConnected(ComponentName className, IBinder rawBinder) {
      Log.d("CONNECT", "ON SERVICE CONNECTED");
      mNokeService = ((NokeDeviceManagerService.LocalBinder) rawBinder).getService();
      mNokeService.registerNokeListener(mNokeServiceListener);
      String message = "On service connected";

      if (!mNokeService.initialize()) {
        Log.e(TAG, "Unable to initialize Bluetooth");
        message = "Unable to initialize Bluetooth";
      }
      final WritableMap event = Arguments.createMap();
      event.putString("message", message);
      event.putBoolean("status", true);
      emitDeviceEvent("onServiceConnected", event);
    }

    public void onServiceDisconnected(ComponentName classname) {
      mNokeService = null;
      final WritableMap event = Arguments.createMap();
      event.putBoolean("status", true);
      emitDeviceEvent("onServiceDisconnected", event);
    }
  };

  private NokeServiceListener mNokeServiceListener = new NokeServiceListener() {
    @Override
    public void onNokeDiscovered(NokeDevice noke) {
      mNokeService.stopScanning();
      if(currentNoke.getMac().equals(noke.getMac())) {
        mNokeService.connectToNoke(noke);
      }
      lastEventCode = 0;
      emitDeviceEvent("onNokeDiscovered", createCommonEvents(noke));
    }

    @Override
    public void onNokeConnecting(NokeDevice noke) {
      emitDeviceEvent("onNokeConnecting", createCommonEvents(noke));
      lastEventCode = 1;
    }

    @Override
    public void onNokeConnected(NokeDevice noke) {
      emitDeviceEvent("onNokeConnected", createCommonEvents(noke));
      lastEventCode = 2;
    }

    @Override
    public void onNokeSyncing(NokeDevice noke) {
      emitDeviceEvent("onNokeSyncing", createCommonEvents(noke));
      lastEventCode = 3;
    }

    @Override
    public void onNokeUnlocked(NokeDevice noke) {
      emitDeviceEvent("onNokeUnlocked", createCommonEvents(noke));
      lastEventCode = 4;
    }

    @Override
    public void onNokeDisconnected(NokeDevice noke) {
      emitDeviceEvent("onNokeDisconnected", createCommonEvents(noke));
      mNokeService.stopScanning();
      lastEventCode = 5;
      mNokeService.uploadData();
    }

    @Override
    public void onDataUploaded(int i, String s) {
        Log.d("UPLOADED", s);
    }

    @Override
    public void onBluetoothStatusChanged(int bluetoothStatus) {
      final WritableMap event = Arguments.createMap();
      event.putInt("code", bluetoothStatus);
      emitDeviceEvent("onBluetoothStatusChanged", event);
    }

    @Override
    public void onError(NokeDevice noke, int error, String message) {
      Log.e(TAG, "NOKE SERVICE ERROR " + error + ": " + message);
//      switch (error) {
//        case NokeMobileError.DEVICE_ERROR_INVALID_KEY:
//          return;
//      }
      final WritableMap event = Arguments.createMap();
      event.putInt("code", error);
      event.putString("message", message);
      lastEventCode = 6;
      emitDeviceEvent("onError", event);
    }
  };

  private static void emitDeviceEvent(String eventName, @Nullable WritableMap eventData) {
    // A method for emitting from the native side to JS
    // https://facebook.github.io/react-native/docs/native-modules-android.html#sending-events-to-javascript
    reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit(eventName, eventData);
  }
}

File: android/src/main/java/com/lynkxyz/noke/RNNokeModule.kt
//  Created by react-native-create-bridge

package com.lynkxyz.noke

import android.bluetooth.BluetoothAdapter
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Handler
import android.os.IBinder
import android.os.Looper
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.noke.nokemobilelibrary.*

class RNNokeModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    private var mNokeService: NokeDeviceManagerService? = null
    private var currentNoke: NokeDevice? = null

    init {
        Companion.reactContext = reactContext
    }

    private val mServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(className: ComponentName, rawBinder: IBinder) {
//            Log.w(TAG, "ON SERVICE CONNECTED")

            //Store reference to service
            mNokeService = (rawBinder as NokeDeviceManagerService.LocalBinder).getService(NokeDefines.NOKE_LIBRARY_SANDBOX)

            //Register callback listener
            mNokeService!!.registerNokeListener(mNokeServiceListener)

            //Start bluetooth scanning
            mNokeService!!.startScanningForNokeDevices()

            if (!mNokeService!!.initialize()) {
//                Log.e(TAG, "Unable to initialize Bluetooth")
                emitDeviceEvent("onServiceConnected", writableMapOf(
                        "connected" to false
                ))
            } else {
                emitDeviceEvent("onServiceConnected", writableMapOf(
                        "connected" to true
                ))
            }
        }

        override fun onServiceDisconnected(classname: ComponentName) {
            mNokeService = null
            emitDeviceEvent("onServiceDisconnected", writableMapOf(
                    "connected" to false
            ))
        }
    }


    private val mNokeServiceListener = object : NokeServiceListener {
        override fun onNokeDiscovered(noke: NokeDevice) {
            currentNoke = noke
            mNokeService!!.connectToNoke(currentNoke)
            emitDeviceEvent("onNokeDiscovered", nokeDeviceInfo(noke))
        }

        override fun onNokeConnecting(noke: NokeDevice) {
            emitDeviceEvent("onNokeConnecting", nokeDeviceInfo(noke))
        }

        override fun onNokeConnected(noke: NokeDevice) {
            currentNoke = noke
            mNokeService!!.stopScanning()
            emitDeviceEvent("onNokeConnected", nokeDeviceInfo(noke))
        }

        override fun onNokeSyncing(noke: NokeDevice) {
            emitDeviceEvent("onNokeSyncing", nokeDeviceInfo(noke))
        }

        override fun onNokeUnlocked(noke: NokeDevice) {
            emitDeviceEvent("onNokeUnlocked", nokeDeviceInfo(noke))
        }

        override fun onNokeShutdown(noke: NokeDevice, isLocked: Boolean?, didTimeout: Boolean?) {
            emitDeviceEvent("onNokeShutdown", writableMapOf(
                    "noke" to nokeDeviceInfo(noke),
                    "isLocked" to isLocked,
                    "didTimeout" to didTimeout
            ))
        }

        override fun onNokeDisconnected(noke: NokeDevice) {
            mNokeService!!.startScanningForNokeDevices()
            mNokeService!!.setBluetoothScanDuration(8000)
            emitDeviceEvent("onNokeDisconnected", nokeDeviceInfo(noke))
        }

        override fun onDataUploaded(result: Int, message: String) {
            emitDeviceEvent("onDataUploaded", writableMapOf(
                    "result" to result,
                    "message" to message
            ))
        }

        override fun onBluetoothStatusChanged(bluetoothStatus: Int) {
            when(bluetoothStatus) {
                BluetoothAdapter.STATE_ON  -> startScan()
            }
            emitDeviceEvent("onBluetoothStatusChanged", writableMapOf(
                    "status" to bluetoothStatus
            ))

        }

        override fun onError(noke: NokeDevice?, error: Int, message: String) {
            emitDeviceEvent("onError", writableMapOf(
                    "noke" to nokeDeviceInfo(noke),
                    "message" to message,
                    "error" to error
                    ))
        }
    }

    override fun getName(): String {
        // Tell React the name of the module
        // https://facebook.github.io/react-native/docs/native-components-android.html#1-create-the-viewmanager-subclass
        return REACT_CLASS
    }

    override fun getConstants(): Map<String, Any>? {
        // Export any constants to be used in your native module
        // https://facebook.github.io/react-native/docs/native-modules-android.html#the-toast-module

        return mapOf<String, Any>(
                "NOKE_LOCK_STATE_LOCKED" to NokeDefines.NOKE_LOCK_STATE_LOCKED,
                "NOKE_LOCK_STATE_UNLOCKED" to NokeDefines.NOKE_LOCK_STATE_UNLOCKED,
                "NOKE_LOCK_STATE_UNSHACKLED" to NokeDefines.NOKE_LOCK_STATE_UNSHACKLED,
                "NOKE_LOCK_STATE_UNKNOWN" to NokeDefines.NOKE_LOCK_STATE_UNKNOWN,

                "ERROR_LOCATION_PERMISSIONS_NEEDED" to NokeMobileError.ERROR_LOCATION_PERMISSIONS_NEEDED,
                "ERROR_LOCATION_SERVICES_DISABLED" to NokeMobileError.ERROR_LOCATION_SERVICES_DISABLED,
                "ERROR_BLUETOOTH_DISABLED" to NokeMobileError.ERROR_BLUETOOTH_DISABLED,
                "ERROR_BLUETOOTH_GATT" to NokeMobileError.ERROR_BLUETOOTH_GATT,
                "DEVICE_ERROR_INVALID_KEY" to NokeMobileError.DEVICE_ERROR_INVALID_KEY
        )
    }

    @ReactMethod
    fun initService() {
        val nokeServiceIntent = Intent(reactApplicationContext, NokeDeviceManagerService::class.java)
        reactApplicationContext.bindService(nokeServiceIntent, mServiceConnection, Context.BIND_AUTO_CREATE)
    }

    @ReactMethod
    fun unlock(commands: ReadableArray, promise: Promise) {
        if (currentNoke == null) {
            promise.reject(RNNokeError.NO_LOCK_CONNECTED, RNNokeError.NO_LOCK_CONNECTED_MESSAGE)
            return
        } else {
            for (i in 0 until commands.size()) {
                currentNoke!!.sendCommands(commands.getString(i))
            }
            promise.resolve(nokeDeviceInfo(currentNoke))
        }
    }

    @ReactMethod
    fun unlockOffline(key: String, command: String, promise: Promise) {
        if (currentNoke == null) {
            promise.reject(RNNokeError.NO_LOCK_CONNECTED, RNNokeError.NO_LOCK_CONNECTED_MESSAGE)
            return
        }

        if (currentNoke!!.session == null) {
            promise.reject(RNNokeError.NO_LOCK_SESSION, RNNokeError.NO_LOCK_SESSION_MESSAGE)
            return
        }

        currentNoke!!.offlineKey = key
        currentNoke!!.offlineUnlockCmd = command
        currentNoke!!.offlineUnlock()
        promise.resolve(nokeDeviceInfo(currentNoke))
    }

    @ReactMethod
    fun changeLock(mac: String, promise: Promise) {
        val noke = NokeDevice(mac, mac)
        if (currentNoke != null) {
            mNokeService!!.disconnectNoke(this.currentNoke!!)
            mNokeService!!.startScanningForNokeDevices()
        }
        mNokeService!!.removeAllNoke()
        mNokeService!!.addNokeDevice(noke)
        currentNoke = noke
        promise.resolve(nokeDeviceInfo(noke))
    }

    @ReactMethod
    fun removeAllLock() {
        mNokeService!!.removeAllNoke()
        mNokeService!!.stopScanning()
    }

    @ReactMethod
    fun startScan() {
        mNokeService!!.startScanningForNokeDevices()
    }

    @ReactMethod
    fun stopScan() {
        mNokeService!!.stopScanning()
    }

    @ReactMethod
    fun disconnect() {
        if(currentNoke != null) {
            mNokeService!!.disconnectNoke(currentNoke)
        }
    }

    @ReactMethod
    fun deviceInfo(promise: Promise) {
        if (currentNoke == null) {
            promise.reject(RNNokeError.NO_LOCK_CONNECTED, RNNokeError.NO_LOCK_CONNECTED_MESSAGE)
        } else {
            promise.resolve(nokeDeviceInfo(currentNoke))
        }
    }

    private fun nokeDeviceInfo(noke: NokeDevice?): WritableMap {
        return if (noke == null) {
            writableMapOf()
        } else {
            writableMapOf(
                    "battery" to noke.battery?.toString(),
                    "connectionState" to noke.connectionState.toString(),
                    "offlineKey" to noke.offlineKey,
                    "lastSeen" to noke.lastSeen.toString(),
                    "lockState" to noke.lockState.toString(),
                    "mac" to noke.mac,
                    "name" to noke.name,
                    "serial" to noke.serial,
                    "session" to noke.session,
                    "trackingKey" to noke.trackingKey?.toString(),
                    "version" to noke.version
//                    "hardwareVersion" to noke.hardwareVersion?.toString(),
//                    "softwareVersion" to noke.softwareVersion?.toString(),
            )
        }
    }

    companion object {
        const val REACT_CLASS = "RNNoke"
        private var reactContext: ReactApplicationContext? = null

        private fun emitDeviceEvent(eventName: String, eventData: WritableMap?) {
            // A method for emitting from the native side to JS
            // https://facebook.github.io/react-native/docs/native-modules-android.html#sending-events-to-javascript
            reactContext!!.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java).emit(eventName, eventData)
        }
    }
}
